### ISSUE TITLE
BUG: Repeated calls to the var method of complex MaskedArrays lead to increasing large results

### ISSUE THREAD

**Issue by RonaldAJ** [numpy/numpy#23896]
### Describe the issue:

On windows10 using the main branch  50984037 (open):Merge pull request #23878 from charris/backport-23846 repeated calls to the var method of a complex MaskedArrays lead to increasingly large results.

An old version on linux does not show the problem:
1.20.3
3.8.10 (default, May 26 2023, 14:05:08) 
[GCC 9.4.0]

​

### Reproduce the code example:

```python
import numpy.ma as ma
import numpy as np

# Base data definition.
x = np.array([8.375j, 7.545j, 8.828j, 8.5j, 1.757j, 5.928,
              8.43, 7.78, 9.865, 5.878, 8.979, 4.732,
              3.012, 6.022, 5.095, 3.116, 5.238, 3.957,
              6.04, 9.63, 7.712, 3.382, 4.489, 6.479j,
              7.189j, 9.645, 5.395, 4.961, 9.894, 2.893,
              7.357, 9.828, 6.272, 3.758, 6.693, 0.993j])

mask = np.array([0, 1, 0, 1, 0, 0,
             1, 0, 1, 1, 0, 1,
             0, 0, 0, 1, 0, 1,
             0, 0, 0, 1, 1, 1,
             1, 0, 0, 1, 0, 0,
             0, 0, 1, 0, 1, 0], dtype=bool)

mx = ma.array(data=x, mask=mask)

mXcomp = mx.compressed()

for i in range(10):
    # print(np.min(mXcomp),np.max(mXcomp))
    #print(mXcomp)
    print(mXcomp.var(), x.var())


'''Python
Example output:
16.211890209804558 19.282807968364196
16.257534793652272 19.282807968364196
16.390019533407436 19.282807968364196
17.271411574130035 19.282807968364196
41.913043714992426 19.282807968364196
13938.461926153328 19.282807968364196
4070437439.6019826 19.282807968364196
3.479376771804534e+20 19.282807968364196
2.5422731712363174e+42 19.282807968364196
1.3572621042094718e+86 19.282807968364196
'''
```

### Runtime information:

1.25.0rc1+90.g50984037f
3.10.11 | packaged by Anaconda, Inc. | (main, Apr 20 2023, 18:56:50) [MSC v.1916 64 bit (AMD64)]

### Context for the issue:

It is a likely pointer/memory leakage problem, and it also affects the outcomes of other computations using the same matrix.

It might be related to the reason some tests for numpy.ma fail:

python runtests.py -t numpy/ma/tests/
is failing with weird errors on complex masked arrays, particularly:
 test_varstd
 test_complex

**Comment by seberg** [comment]
Can't repro and it seems a bit overly strange so I have to ask: Can you double check that this isn't not with your modifications?

**Comment by RonaldAJ** [comment]
Sanity check: origin/main and upstream/main are on the same commit. The one indicated above. So I didn't include my own commits in main.

**Comment by seberg** [comment]
> I now manually removed some build products to be sure of a clean environment, and rebuilt.

The best way to do this is probably `git clean -xdf` to clean up everything in the repository.  Using `runtests` should drop you into the correct thing, so not sure what might be off.

Can you try to check the `test_complex` in isolation and see if you can boil it down further?  It looks a bit closer to a minimally weird example.
(would be nice to factor out `average` to the more base calls also, because I am not quite sure what it does behind the scenes).

Or maybe just try if using `mean` alone works in your first?

In your example:
```
    def test_complex(self):
        # Test with complex data.
        # (Regression test for https://github.com/numpy/numpy/issues/2684)
        mask = np.array([[0, 0, 0, 1, 0],
                         [0, 1, 0, 0, 0]], dtype=bool)
        a = masked_array([[0, 1+2j, 3+4j, 5+6j, 7+8j],
                          [9j, 0+1j, 2+3j, 4+5j, 7+7j]],
                         mask=mask)

        av = average(a)
        expected = np.average(a.compressed())
        assert_almost_equal(av.real, expected.real)
        assert_almost_equal(av.imag, expected.imag)

        av0 = average(a, axis=0)
        expected0 = average(a.real, axis=0) + average(a.imag, axis=0)*1j
        assert_almost_equal(av0.real, expected0.real)
>       assert_almost_equal(av0.imag, expected0.imag)

a          = masked_array(
  data=[[0j, (1+2j), (3+4j), --, (7+8j)],
        [9j, --, (2+3j), (4+5j), (7+7j)]],
  mask=[[False, False, False,  True, False],
        [False,  True, False, False, False]],
  fill_value=(1e+20+0j))
av         = (3+4.75j)
av0        = masked_array(data=[4.5j, (1+2j), (2.5+3.5j), (4+5j),
                   (3.182993687e-314+2j)],
             mask=[False, False, False, False, False],
       fill_value=(1e+20+0j))
expected   = (3+4.75j)
expected0  = masked_array(data=[4.5j, (1+2j), (2.5+3.5j), (4+5j),
                   (6.365987374e-314+4j)],
             mask=[False, False, False, False, False],
       fill_value=(1e+20+0j))
mask       = array([[False, False, False,  True, False],
       [False,  True, False, False, False]])
self       = <numpy.ma.tests.test_extras.TestAverage object at 0x000001FD7EF28F10>
```
`av0.real` and `expected0.real` are asserted to be equal, but then fail the `imag` equality check with *both* finally containing garbage in the last element, but the garbage in `expected0` is twice the garbage of that in `av0`...

`assert_almost_equal` is rather complicated unfortunately, so see if `assert (arr == arr2).all()` would be better to be minimal.

Right now, my only good idea is to try and narrow things down a bit more.  Maybe also post `np.show_runtime()`.  I don't quite expect it, but maybe it makes sense to check adding `NPY_DISABLE_OPTIMIZATION=1`  (I have to search, I am not sure `set CFLAGS=/D NPY_DISABLE_OPTIMIZATION=1` works on windows)

**Comment by RonaldAJ** [comment]
Observation: 
Complex subtraction of a constantant from an array shows of by one behavior.  Last result in the output array is bogus and depends on the actual run.

```Python
import numpy.ma as ma
import numpy as np

# Base data definition.
x = np.array([8.375j, 7.545j, 8.828j, 8.5j, 1.757j, 5.928,
              8.43, 7.78, 9.865, 5.878, 8.979, 4.732,
              3.012, 6.022, 5.095, 3.116, 5.238, 3.957,
              6.04, 9.63, 7.712, 3.382, 4.489, 6.479j,
              7.189j, 9.645, 5.395, 4.961, 9.894, 2.893,
              7.357, 9.828, 6.272, 3.758, 6.693, 0.993j])

m = np.array([0, 1, 0, 1, 0, 0,
             1, 0, 1, 1, 0, 1,
             0, 0, 0, 1, 0, 1,
             0, 0, 0, 1, 1, 1,
             1, 0, 0, 1, 0, 0,
             0, 0, 1, 0, 1, 0], dtype=bool)

value_list = [value for mask, value in zip(m.tolist(), x.tolist()) if  not mask]
values = np.array(value_list)

my_mean = np.sum(values)/len(values)
my_diff = values - my_mean

_diff_conj = np.conjugate(my_diff)
my_var = np.sum((my_diff*_diff_conj).real)/len(values)

values[-1]-my_mean, my_diff
```

Output:
```
((-5.438380952380951+0.04285714285714293j),
 array([-5.43838095+7.42485714j, -5.43838095+7.87785714j,
        -5.43838095+0.80685714j,  0.48961905-0.95014286j,
         2.34161905-0.95014286j,  3.54061905-0.95014286j,
        -2.42638095-0.95014286j,  0.58361905-0.95014286j,
        -0.34338095-0.95014286j, -0.20038095-0.95014286j,
         0.60161905-0.95014286j,  4.19161905-0.95014286j,
         2.27361905-0.95014286j,  4.20661905-0.95014286j,
        -0.04338095-0.95014286j,  4.45561905-0.95014286j,
        -2.54538095-0.95014286j,  1.91861905-0.95014286j,
         4.38961905-0.95014286j, -1.68038095-0.95014286j,
         0.        +0.993j     ]))
 ```

Last entry in the array should correspond with the value at the top, but it doesn't. It also varies with each run.

**Comment by RonaldAJ** [comment]
> Just to be sure, you are using a typical MSVC toolchain and not some mingw or so?

I am not sure. Build management is not my strongest point. But I am under the impression that I am not using mingw. I see the MSVC compiler being called during setup.

```bash
 C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\bin\HostX86\x64\link.exe 
 C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\bin\HostX86\x64\cl.exe
```

**Comment by seberg** [comment]
Maybe as a start, can you try `SET NPY_DISABLE_CPU_FEATURES="AVX2"` and retry?  And the same with `SET NPY_DISABLE_CPU_FEATURES="SSE41"` (this should disable the use of most SIMD)?
(Sorry, probably not the fasted thing to try, but hoping to get a bit of information without thinking about it too much.)

**Comment by RonaldAJ** [comment]
You might have to be more explicit about how and where to set those compiler flags.

I am working from a git-bash shell under windows. The instructions with 'set' are for the CMD or powershell? 

I can pass --disable-compiler after build_ext, but after build it yields crash.

**Comment by RonaldAJ** [comment]
runtests runs with this option but the output is the same. Same two test failing. 

But judging from the header SSE41 and AVX are still in?

```
$ python runtests.py --disable-optimization -t numpy/ma/tests/
Building, see build.log...
Build OK
C:\GIT_Workspace\numpy\build\testenv\Lib\site-packages\numpy\_pytesttester.py:143: DeprecationWarning:

  `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
  of the deprecation of `distutils` itself. It will be removed for
  Python >= 3.12. For older Python versions it will remain present.
  It is recommended to use `setuptools < 60.0` for those Python versions.
  For more details, see:
    https://numpy.org/devdocs/reference/distutils_status_migration.html


  from numpy.distutils import cpuinfo
'lscpu' is not recognized as an internal or external command,
operable program or batch file.
NumPy version 1.25.0rc1+90.g50984037f
NumPy relaxed strides checking option: True
NumPy CPU features:  SSE SSE2 SSE3 SSSE3* SSE41* POPCNT* SSE42* AVX* F16C* FMA3* AVX2* AVX512F? AVX512CD? AVX512_SKX? AVX512_CLX? AVX512_CNL? AVX512_ICL?

```

**Comment by RonaldAJ** [comment]
Nasty effects have gone now. 

Answers are correct and rerunning doesn't change them. From Jupyter it is also fine.

**Comment by seberg** [comment]
Cool, thanks!  So it has to do with SIMD/optimizations and we also know that the issue is in the trailing elements of an array that has a size that isn't a multiple of the vectorization, I think.

Did you narrow down which function is causing it?  The sum?

**Comment by RonaldAJ** [comment]
I had to do:
```bash
export CFLAGS="NPY_DISABLE_OPTIMIZATION=1 NPY_DISABLE_CPU_FEATURES=\"AVX2\" NPY_DISABLE_CPU_FEATURES=\"SSE41\""
```
to set the compile flags in git-bash.

**Comment by RonaldAJ** [comment]
Individual flags:outcome
NPY_DISABLE_CPU_FEATURES="SSE41" : problem exists
NPY_DISABLE_CPU_FEATURES="AVX2": problem exists
NPY_DISABLE_OPTIMIZATION=1: problem is gone

**Comment by seberg** [comment]
Hmmm, the problem should be here (or in its calls):
https://github.com/numpy/numpy/blob/85de7693e35547d0e387bc08ad28734aa51ac44b/numpy/core/src/umath/loops_arithm_fp.dispatch.c.src#LL542-L558

(in the above example, I think the second operand is the scalar, the first should be contiguous, so it is this specific branch).

That code isn't exactly easy to read, but I don't see anything obviously wrong.  Also, though, I don't see why you are seeing it and not me locally (or CI for that matter).

I hyave tried to look at the store/load code a bit, but can't really see anything obvious.  Would be nice to know which specific instruction set `SSE3`(?) is at fault.
(If the disabling at runtime worked, it should be SSE3 I guess.  But really this should all be similar on linux, so I am not sure what the problem could be.  Maybe alignment issues, or a compiler issue even?)

**Comment by RonaldAJ** [comment]
Only providing those flags at runtime doesn't help.

**Comment by seberg** [comment]
Yeah, its probably in the SSE version or even all.  But I am still stumped that you see it on your windows only.  Sometimes there are compiler bugs around it, but I am not quite there yet.

**Comment by seberg** [comment]
@seiko2plus do you have a quick idea here?

**Comment by seberg** [comment]
@rdevulap and @Mousius, sorry to ask:  Do you have a thought what might be wrong here on windows/MSVC?  I am a bit worried that this is a regression leading to incorrect results on some platforms (buggy compiler, ...?)

**Comment by seberg** [comment]
Maybe we should really consider if this is a compiler issue?  I think we are using a Visual Studio 19.20+ version while this is on an older one (not 100% sure MSVC versioning is confusing...).

If that is so, maybe we should just add a guard here and acknowledge that we don't quite know what is wrong (for release purposes):
```
#if defined(_M_X64) && defined(_MSC_VER) && _MSC_VER >= 1920 && !defined(__clang__)
```

@RonaldAJ would you be up for trying whether upgrading visual studio helps?

**Comment by RonaldAJ** [comment]
I will give it a try.

**Comment by Mousius** [comment]
@seberg I can't see anything obviously wrong with the code but I've just installed a completely fresh set of stuff on Windows and it looks like it's occurring for me 😿  will dig a bit further.

**Comment by RonaldAJ** [comment]
No change, as in problem persist:

 C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.3
6.32532\bin\HostX86\x64\cl.exe

**Comment by seberg** [comment]
Nah, I think its a new one for sure now (probably even before).  Adding a guard will only be useful if we know what to guard for (unless we just blanket guard MSVC right now assuming that other compilers are fine).

**Comment by r-devulap** [comment]
Given that it goes away when disabling AVX2 and SSE and the element being off, this looks like a masked load gone wrong. I wonder if you still see the problem if the array size if a multiple of 8?

**Comment by RonaldAJ** [comment]
Running:

```Python
import numpy.ma as ma
import numpy as np
from numpy.testing import assert_almost_equal

# Base data definition.
x = np.array([8.375j, 7.545j, 8.828j, 8.5j, 1.757j, 5.928,
              8.43, 7.78, 9.865, 5.878, 8.979, 4.732,
              3.012, 6.022, 5.095, 3.116, 5.238, 3.957,
              6.04, 9.63, 7.712, 3.382, 4.489, 6.479j,
              7.189j, 9.645, 5.395, 4.961, 9.894, 2.893,
              7.357, 9.828, 6.272, 3.758, 6.693, 0.993j])

value_list=[]

for index, value in enumerate(x.tolist()):

    value_list.append(value) 
    values = np.array(value_list)

    my_mean = np.sum(values)/len(values)
    my_diff = values - my_mean

    _diff_conj = np.conjugate(my_diff)
    my_var = np.sum((my_diff*_diff_conj).real)/len(values)

    #print(index,value, my_mean, my_diff[-1])
    
    try:
        assert_almost_equal(value-my_mean,my_diff[-1])
    except Exception as e:
        print(f'Failed at {index=} {value=}, {my_mean=}, {my_diff[-1]=}')
```

I get a crash at every even length:

```
$ python numpy_bug_23896c.py
$ python numpy_bug_23896c.py
'lscpu' is not recognized as an internal or external command,
operable program or batch file.
Failed at index=2 value=8.828j, my_mean=8.249333333333333j, my_diff[-1]=(1.33509
38861175663e-306+3.594129e-317j)
Failed at index=4 value=1.757j, my_mean=7.0009999999999994j, my_diff[-1]=(9.3460
09626940552e-307+3.6558825666693763e+233j)
Failed at index=6 value=(8.43+0j), my_mean=(2.051142857142857+5.000714285714285j
), my_diff[-1]=(1.3351237635799904e-306+9.346097902587135e-307j)
Failed at index=8 value=(9.865+0j), my_mean=(3.5558888888888887+3.88944444444444
36j), my_diff[-1]=(9.791071931451263e-307+3.3350978e-317j)
Failed at index=10 value=(8.979+0j), my_mean=(4.26+3.182272727272727j), my_diff[
-1]=0j
Failed at index=12 value=(3.012+0j), my_mean=(4.2003076923076925+2.6926923076923
073j), my_diff[-1]=(1.7975706653261772e-152+2.0264744118034976e+267j)
Failed at index=14 value=(5.095+0j), my_mean=(4.3814+2.3336666666666663j), my_di
ff[-1]=0j
Failed at index=16 value=(5.238+0j), my_mean=(4.357352941176471+2.05911764705882
3j), my_diff[-1]=(1.226143386058e-311+1.2261407449175e-311j)
Failed at index=18 value=(6.04+0j), my_mean=(4.424842105263158+1.842368421052631
2j), my_diff[-1]=(1.0681135425336275e-306+8.344442887376749e-308j)
Failed at index=20 value=(7.712+0j), my_mean=(4.829238095238095+1.66690476190476
17j), my_diff[-1]=(1.226160664751e-311+1.2261606878496e-311j)
Failed at index=22 value=(4.489+0j), my_mean=(4.751521739130435+1.52195652173913
02j), my_diff[-1]=(6.952565715854e-310+6.95256571182477e-310j)
Failed at index=24 value=7.189j, my_mean=(4.3714+1.9469199999999998j), my_diff[-
1]=(8.364707998415743e-299+1.0159267464276398e-298j)
Failed at index=26 value=(5.395+0j), my_mean=(4.6046296296296285+1.8027037037037
035j), my_diff[-1]=(5.13078772946714e-294+1.2855628473482465e-303j)
Failed at index=28 value=(9.894+0j), my_mean=(4.799310344827586+1.67837931034482
77j), my_diff[-1]=(1.2261456554715e-311+1.2261456558193e-311j)
Failed at index=30 value=(7.357+0j), my_mean=(4.820322580645161+1.57009677419354
83j), my_diff[-1]=(7.904251805221914e-120+5.494708848813239e-96j)
Failed at index=32 value=(6.272+0j), my_mean=(5.0160606060606066+1.4749393939393
94j), my_diff[-1]=0j
Failed at index=34 value=(6.693+0j), my_mean=(5.028028571428572+1.39065714285714
28j), my_diff[-1]=0j
```

**Comment by RonaldAJ** [comment]
@r-devulap 

> Given that it goes away when disabling AVX2 and SSE and the last element being off, this looks like a masked load gone wrong. I wonder if you still see the problem if the array size if a multiple of 8?

Does this [link](https://lemire.me/blog/2022/11/08/modern-vector-programming-with-masked-loads-and-stores/ )  describe the phenomenon you mentioned?

**Comment by RonaldAJ** [comment]
With export CFLAGS="NPY_DISABLE_OPTIMIZATION=1" the script raises no exceptions.

**Comment by RonaldAJ** [comment]
@seberg I have been porting octave code all day. I have been off by one many times today.

**Comment by r-devulap** [comment]
> I get an exception at every even index except 0:

Interesting that index 0 (len = 1) doesn't hit the bug. It skips the for loop and directly uses the masked load/store. All other even indices go through for loop and then hit the masked load/store.

**Comment by RonaldAJ** [comment]
As it seems to get quite low level my CPU is a AMD Ryzen 7 5800H with Radeon Graphics.

**Comment by r-devulap** [comment]
Simplest reproducible example: 

```
values = np.array([1.j, 2.j, 3.j])
my_mean = np.sum(values)/len(values)
my_diff = values - my_mean

try:
    assert_almost_equal(values[2]-my_mean,my_diff[2])
except Exception as e:
    print(f'Failed {values[2]-my_mean=}, {my_diff[2]=}')
```
 Output: 

`Failed values[2]-my_mean=1j, my_diff[2]=(1.3350938861175663e-306+2.78134235907529e-309j)`

Still debugging to find the root cause.

**Comment by r-devulap** [comment]
I think the culprit is this line: https://github.com/numpy/numpy/blob/85de7693e35547d0e387bc08ad28734aa51ac44b/numpy/core/src/umath/loops_arithm_fp.dispatch.c.src#L550

This is suppose to process the tail end of the array which is supposed to be shorter than the register size. So i am not sure why it is a for loop.

**Comment by r-devulap** [comment]
@mattip Complex add, subtract and multiply with a scalar are all broken: It somehow works on linux, but i think that is buggy too. Simplest reproducible example: 

```
values = np.array([1.j, 2.j, 3.j])
print(values + 1.)
```
Output: (clearly the last value is garbage). 

`[1.00000000e+000+1.00000000e+000j 1.00000000e+000+2.00000000e+000j, 9.34600963e-307+5.56268465e-309j]`

**Comment by charris** [comment]
A fix for this will need a backport, correct?

**Comment by r-devulap** [comment]
yeah, it will. Basic complex math is broken, so we will need to fix it. Still trying to understand the code and working on a fix.

**Comment by charris** [comment]
OK, I'll wait to do the final release of 1.25.0.

**Comment by r-devulap** [comment]
Okay, so i was wrong about the for loop. I now understand why that's there and linux version seems to be fine. The bug is definitely coming from this line: 

https://github.com/numpy/numpy/blob/85de7693e35547d0e387bc08ad28734aa51ac44b/numpy/core/src/umath/loops_arithm_fp.dispatch.c.src#L557

`npyv_store2_till_@sfx@(dst, len, r);` when `len ==1 `

The implementation of `npyv_store2_till_f64` is a bit hazy to me. After back tracking a lot of macros, I think it goes from `npyv_store2_till_s64 `to `npyv_storel_s64 `to https://github.com/numpy/numpy/blob/85de7693e35547d0e387bc08ad28734aa51ac44b/numpy/core/src/common/simd/avx2/memory.h#L28

I am still not sure if this is a compiler bug or a code bug. But debugging this has been hard, might need @seiko2plus help.

**Comment by mattip** [comment]
Can you get the assembler produced on windows and compare it to linux?

**Comment by mattip** [comment]
Does this happen with the wheel from https://anaconda.org/scientific-python-nightly-wheels/numpy/ ?

**Comment by mattip** [comment]
I cannot reproduce on my windows machine on ryzen:
```
>>> import numpy as np
>>> values = np.array([1.j, 2.j, 3.j])
>>> print(values + 1.0)
[1.+1.j 1.+2.j 1.+3.j]
>>> np.show_runtime()
[{'numpy_version': '1.25.0rc1',
  'python': '3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 '
            '64 bit (AMD64)]',
  'uname': uname_result(system='Windows', node='desktop-ryzen', release='10', version='10.0.19045', machine='AMD64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}}]
>>> np.__version__()
'1.25.0rc1'
```

**Comment by mattip** [comment]
(I am using Visual Studio Community 2019)

**Comment by r-devulap** [comment]
huh, that's odd. I am using `MSC v.1934`. BTW, I just noted an interesting behavior. Could you run `print(arr + 1.0) `twice back to back? It fails 1 out of two times. It doesn't run the SIMD code everytime (which is really odd), but when it does it gives the wrong answer: (I have a lot of DEBUG info printing in the SIMD code))  

```
>>> arr = np.array([1.0j, 2.0j, 3.0j])
>>> arr + 1
array([1.+1.j, 1.+2.j, 1.+3.j])
>>> arr + 1.
RD_DEBUG: len = 3
RD_DEBUG: vstep = 4, wstep = 8, hstep = 2
RD_DEBUG: masked load, len = 3. src = 0.000000, 1.000000, 0.000000, 2.000000, 0.000000, 3.000000
arr = 0.000000, 1.000000, 0.000000, 2.000000
arr = 1.000000, 1.000000, 1.000000, 2.000000
dst = 1.000000, 1.000000, 1.000000, 2.000000
RD_DEBUG: masked load, len = 1. src = 0.000000, 3.000000, 1.000000, 1.000000, 1.000000, 2.000000
arr = 0.000000, 3.000000, 0.000000, 0.000000
arr = 1.000000, 3.000000, 1.000000, 0.000000
dst = 0.000000, 0.000000, -0.000000, 0.000000
array([1.+1.j, 1.+2.j, 0.+0.j])
>>> arr + 1.
array([1.+1.j, 1.+2.j, 1.+3.j])
>>> arr + 1.
RD_DEBUG: len = 3
RD_DEBUG: vstep = 4, wstep = 8, hstep = 2
RD_DEBUG: masked load, len = 3. src = 0.000000, 1.000000, 0.000000, 2.000000, 0.000000, 3.000000
arr = 0.000000, 1.000000, 0.000000, 2.000000
arr = 1.000000, 1.000000, 1.000000, 2.000000
dst = 1.000000, 1.000000, 1.000000, 2.000000
RD_DEBUG: masked load, len = 1. src = 0.000000, 3.000000, 1.000000, 1.000000, 1.000000, 2.000000
arr = 0.000000, 3.000000, 0.000000, 0.000000
arr = 1.000000, 3.000000, 1.000000, 0.000000
dst = 0.000000, 0.000000, -0.000000, 0.000000
array([1.+1.j, 1.+2.j, 0.+0.j])
>>> arr + 1.
array([1.+1.j, 1.+2.j, 1.+3.j])
>>> arr + 1.
RD_DEBUG: len = 3
RD_DEBUG: vstep = 4, wstep = 8, hstep = 2
RD_DEBUG: masked load, len = 3. src = 0.000000, 1.000000, 0.000000, 2.000000, 0.000000, 3.000000
arr = 0.000000, 1.000000, 0.000000, 2.000000
arr = 1.000000, 1.000000, 1.000000, 2.000000
dst = 1.000000, 1.000000, 1.000000, 2.000000
RD_DEBUG: masked load, len = 1. src = 0.000000, 3.000000, 1.000000, 1.000000, 1.000000, 2.000000
arr = 0.000000, 3.000000, 0.000000, 0.000000
arr = 1.000000, 3.000000, 1.000000, 0.000000
dst = 0.000000, 0.000000, -0.000000, 0.000000
array([1.+1.j, 1.+2.j, 0.+0.j])
```

**Comment by r-devulap** [comment]
> Can you get the assembler produced on windows and compare it to linux?

As soon as I can figure out how to do that in windows (this is my first time debugging low level code with visual studio :( )

**Comment by seiko2plus** [comment]
Sorry for the delayed response, I can confirm its definitely a compiler bug since msvc v19.30, and it can also be trrigerd via `test_simd.py`:
```Bash

self = <numpy.core.tests.test_simd.Test_SIMD_ALL_256_AVX2_f64 object at 0x0000022A2FE26FD0>, intrin = 'self.store2_till', elsizes = (32, 64), scale = 2

    @pytest.mark.parametrize("intrin, elsizes, scale", [
        ("self.store_till", (32, 64), 1),
        ("self.store2_till", (32, 64), 2),
    ])
    def test_memory_partial_store(self, intrin, elsizes, scale):
        if self._scalar_size() not in elsizes:
            return
        npyv_store_till = eval(intrin)
        data = self._data()
        data_rev = self._data(reverse=True)
        vdata = self.load(data)
        lanes = list(range(1, self.nlanes + 1))
        lanes += [self.nlanes**2, self.nlanes**4]
        for n in lanes:
            data_till = data_rev.copy()
            data_till[:n*scale] = data[:n*scale]
            store_till = self._data(reverse=True)
            npyv_store_till(store_till, n, vdata)
>           assert store_till == data_till
E           assert [4.0, 3.0, 2.0, 1.0] == [1.0, 2.0, 2.0, 1.0]
E             At index 0 diff: 4.0 != 1.0
E             Full diff:
E             - [1.0, 2.0, 2.0, 1.0]
E             + [4.0, 3.0, 2.0, 1.0]

data       = [1.0, 2.0, 3.0, 4.0]
data_rev   = [4.0, 3.0, 2.0, 1.0]
data_till  = [1.0, 2.0, 2.0, 1.0]
elsizes    = (32, 64)
intrin     = 'self.store2_till'
lanes      = [1, 2, 3, 4, 16, 256]
n          = 1
npyv_store_till = <built-in function store2_till_f64>
scale      = 2
self       = <numpy.core.tests.test_simd.Test_SIMD_ALL_256_AVX2_f64 object at 0x0000022A2FE26FD0>
store_till = [4.0, 3.0, 2.0, 1.0]
vdata      = <npyv_f64 of [1.0, 2.0, 3.0, 4.0]>

..\..\..\AppData\Roaming\Python\Python311\site-packages\numpy-1.25.0rc1+189.gff5c1bf38-py3.11-win-amd64.egg\numpy\core\tests\test_simd.py:727: AssertionError
=========================================================================================== short test summary info ============================================================================================
FAILED ..\..\core\tests\test_simd.py::Test_SIMD_ALL_256_AVX2_u64::test_memory_partial_store[self.store2_till-elsizes1-2] - assert [4, 3, 2, 1] == [1, 2, 2, 1]
FAILED ..\..\core\tests\test_simd.py::Test_SIMD_ALL_256_AVX2_s64::test_memory_partial_store[self.store2_till-elsizes1-2] - assert [4, 3, 2, 1] == [1, 2, 2, 1]
FAILED ..\..\core\tests\test_simd.py::Test_SIMD_ALL_256_AVX2_f64::test_memory_partial_store[self.store2_till-elsizes1-2] - assert [4.0, 3.0, 2.0, 1.0] == [1.0, 2.0, 2.0, 1.0]
```

Only AVX2 is affected by this bug, so disabling `AVX2` via `NPY_DISABLE_CPU_FEATURES` should bypass it.

Here a proof of the bug https://godbolt.org/z/4GPP4KT4d, simply the compiler optimizer performs 256-bit mov rather than 128-bit when the pair length is `1`:
```ASM
npyv_store2_till_s64 PROC                         ; COMDAT
        vmovdqu ymm0, YMMWORD PTR [r8]
        cmp     rdx, 1
        jbe     SHORT $LN6@npyv_store
        vmovdqu YMMWORD PTR [rcx], ymm0
$LN6@npyv_store:
        vzeroupper
        ret     0
```
Expected to be:

```Asm
        vmovdqu ymm0, YMMWORD PTR [r8]
        vmovdqu XMMWORD PTR [rcx], xmm0
        cmp     rdx, 1
        jbe     SHORT $LN6@npyv_store
        vextracti128 XMMWORD PTR [rcx+16], ymm0, 1
$LN6@npyv_store:
```

**Comment by charris** [comment]
Thanks Sayed, that will definitely save us time and trouble.

**Comment by seiko2plus** [comment]
#23953 should fix this issue

**Comment by seiko2plus** [comment]
@RonaldAJ,

> Individual flags:outcome
> problem exists NPY_DISABLE_CPU_FEATURES="AVX2":

Are you sure about it? It seems to me that disabling `AVX2` can bypass it.

Would you please define the env var `NPY_DISABLE_CPU_FEATURES="AVX2"`,
and check numpy config to see if `AVX2` is marked as "not_found" or still part of the "found" list?

**Comment by r-devulap** [comment]
Thanks@seiko2plus, for putting an end to my misery :)

**Comment by mattip** [comment]
Has this been reported to MSVC? I tried [seraching their issue forum](https://developercommunity.visualstudio.com/VisualStudio?ftype=problem&q=compiler+reordering+avx256+vmovdqu) (login required?) but perhaps my search foo is not up to the task. A new issue can be reported via https://developercommunity.visualstudio.com/cpp/report , the godbolt link will probably be enough to demonstrate the issue.

**Comment by mattip** [comment]
While I don't doubt the fix, could someone with that compiler test that the test fails before, passes after #23953? I guess one conclusion is that we should be using the tests more to help us debug this kind of failure.

**Comment by RonaldAJ** [comment]
> Are you sure about it? It seems to me that disabling `AVX2` can bypass it.
> 
> Would you please define the env var `NPY_DISABLE_CPU_FEATURES="AVX2"`, and check numpy config to see if `AVX2` is marked as "not_found" or still part of the "found" list?

Sorry I was passing it in via CFLAGS. If I set it as an environment variable it works as you expect it:  

```
Ronald@Tachyon MINGW64 /c/GIT_Workspace/vingeroefeningen
$  export NPY_DISABLE_CPU_FEATURES=AVX2
(numpy_test)
Ronald@Tachyon MINGW64 /c/GIT_Workspace/vingeroefeningen
$ python numpy_bug_23896c.py
'lscpu' is not recognized as an internal or external command,
operable program or batch file.
(numpy_test)
Ronald@Tachyon MINGW64 /c/GIT_Workspace/vingeroefeningen
$  export NPY_DISABLE_CPU_FEATURES=SSE41
(numpy_test)
Ronald@Tachyon MINGW64 /c/GIT_Workspace/vingeroefeningen
$ python numpy_bug_23896c.py
'lscpu' is not recognized as an internal or external command,
operable program or batch file.
Failed at index=2 value=8.828j, my_mean=8.249333333333333j, my_diff[-1]=1.937e-321j
Failed at index=4 value=1.757j, my_mean=7.0009999999999994j, my_diff[-1]=(9.346009626940552e-307+2.
6379118833184442e+185j)
Failed at index=6 value=(8.43+0j), my_mean=(2.051142857142857+5.000714285714285j), my_diff[-1]=(1.2
906134589073559e-306+1.4241097430522645e-306j)
Failed at index=8 value=(9.865+0j), my_mean=(3.5558888888888887+3.8894444444444436j), my_diff[-1]=(
1.3351067878519732e-306+2.2252259654242998e-306j)
Failed at index=10 value=(8.979+0j), my_mean=(4.26+3.182272727272727j), my_diff[-1]=0j
Failed at index=12 value=(3.012+0j), my_mean=(4.2003076923076925+2.6926923076923073j), my_diff[-1]=
(1.1009987272743e-311+1.1009987270727e-311j)
Failed at index=14 value=(5.095+0j), my_mean=(4.3814+2.3336666666666663j), my_diff[-1]=(2.225225964
1187532e-306+1.1202264247238343e+243j)
Failed at index=16 value=(5.238+0j), my_mean=(4.357352941176471+2.059117647058823j), my_diff[-1]=(1
.1009819669194e-311+1.100979454125e-311j)
Failed at index=18 value=(6.04+0j), my_mean=(4.424842105263158+1.8423684210526312j), my_diff[-1]=(1
.0681135425336275e-306+8.344442887376749e-308j)
Failed at index=20 value=(7.712+0j), my_mean=(4.829238095238095+1.6669047619047617j), my_diff[-1]=(
1.100979785813e-311+1.100977242324e-311j)
Failed at index=22 value=(4.489+0j), my_mean=(4.751521739130435+1.5219565217391302j), my_diff[-1]=(
1.100975618492e-311+1.100975616927e-311j)
Failed at index=24 value=7.189j, my_mean=(4.3714+1.9469199999999998j), my_diff[-1]=(1.10097563299e-
311+1.100975618429e-311j)
Failed at index=26 value=(5.395+0j), my_mean=(4.6046296296296285+1.8027037037037035j), my_diff[-1]=
0j
Failed at index=28 value=(9.894+0j), my_mean=(4.799310344827586+1.6783793103448277j), my_diff[-1]=(
1.1010006274073e-311+1.101000627423e-311j)
Failed at index=30 value=(7.357+0j), my_mean=(4.820322580645161+1.5700967741935483j), my_diff[-1]=(
1.0105468554382593e+261+9.267232631864497e+242j)
Failed at index=32 value=(6.272+0j), my_mean=(5.0160606060606066+1.474939393939394j), my_diff[-1]=0
j
Failed at index=34 value=(6.693+0j), my_mean=(5.028028571428572+1.3906571428571428j), my_diff[-1]=0
j

```

**Comment by RonaldAJ** [comment]
@mattip I'll try.

**Comment by RonaldAJ** [comment]
For me the fix works:

```
Ronald@Tachyon MINGW64 /c/GIT_Workspace/vingeroefeningen
$  export NPY_DISABLE_CPU_FEATURES=""
(numpy_test)
Ronald@Tachyon MINGW64 /c/GIT_Workspace/vingeroefeningen
$ python numpy_bug_23896c.py
'lscpu' is not recognized as an internal or external command,
operable program or batch file.
(numpy_test)
Ronald@Tachyon MINGW64 /c/GIT_Workspace/vingeroefeningen
$
```

Last part build info and git information for completeness:

```
########### EXT COMPILER OPTIMIZATION ###########
INFO: Platform      :
  Architecture: x64
  Compiler    : msvc

CPU baseline  :
  Requested   : 'min'
  Enabled     : SSE SSE2 SSE3
  Flags       : none
  Extra checks: none

CPU dispatch  :
  Requested   : 'max -xop -fma4'
  Enabled     : SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   :
              :
  SSE41       : SSE SSE2 SSE3 SSSE3
  Flags       :
  Extra checks: none
  Detect      : SSE SSE2 SSE3 SSSE3 SSE41
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_unary_fp.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_unary_fp_le.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_arithmetic.dispatch.c
              : numpy\core\src\umath\_umath_tests.dispatch.c
              :
  SSE42       : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT
  Flags       :
  Extra checks: none
  Detect      : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42
              : build\src.win-amd64-3.10\numpy\core\src\multiarray\argfunc.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_comparison.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\_simd\_simd.dispatch.c
              :
  AVX2        : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3
  Flags       : /arch:AVX2
  Extra checks: none
  Detect      : AVX F16C FMA3 AVX2
              : build\src.win-amd64-3.10\numpy\core\src\multiarray\argfunc.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_unary.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_arithm_fp.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_arithmetic.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_logical.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_minmax.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_trigonometric.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_exponent_log.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_hyperbolic.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_comparison.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_unary_complex.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_autovec.dispatch.c
              : numpy\core\src\umath\_umath_tests.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\_simd\_simd.dispatch.c
              :
  AVX512F     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512CD AVX512_SKX
  Flags       : /arch:AVX512
  Extra checks: AVX512F_REDUCE
  Detect      : AVX512_SKX
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_trigonometric.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_unary_complex.dispatch.c
              :
  AVX512_SKX  : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD
  Flags       : /arch:AVX512
  Extra checks: AVX512BW_MASK AVX512DQ_MASK
  Detect      : AVX512_SKX
              : numpy\core\src\npysort\simd_qsort.dispatch.cpp
              : build\src.win-amd64-3.10\numpy\core\src\multiarray\argfunc.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_unary.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_arithmetic.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_logical.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_minmax.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_umath_fp.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_exponent_log.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_hyperbolic.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\umath\loops_comparison.dispatch.c
              : build\src.win-amd64-3.10\numpy\core\src\_simd\_simd.dispatch.c
              :
  AVX512_ICL  : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_SKX AVX512_CLX AVX512_CNL
  Flags       : /arch:AVX512
  Extra checks: none
  Detect      : AVX512_ICL
              : numpy\core\src\npysort\simd_qsort_16bit.dispatch.cpp
INFO: CCompilerOpt.cache_flush[863] : write cache to path -> C:\GIT_Workspace\numpy\build\temp.win-amd64-cpython-310\Release\ccompiler_opt
_cache_ext.py
INFO:
########### CLIB COMPILER OPTIMIZATION ###########
INFO: Platform      :
  Architecture: x64
  Compiler    : msvc

CPU baseline  :
  Requested   : 'min'
  Enabled     : SSE SSE2 SSE3
  Flags       : none
  Extra checks: none

CPU dispatch  :
  Requested   : 'max -xop -fma4'
  Enabled     : SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : none
INFO: CCompilerOpt.cache_flush[863] : write cache to path -> C:\GIT_Workspace\numpy\build\temp.win-amd64-cpython-310\ccompiler_opt_cache_c
lib.py
(numpy_test)
Ronald@Tachyon MINGW64 /c/GIT_Workspace/numpy (avx2_fix)
$ git status
On branch avx2_fix
nothing to commit, working tree clean
(numpy_test)
Ronald@Tachyon MINGW64 /c/GIT_Workspace/numpy (avx2_fix)
$ git rev-parse HEAD
bcc428c3a700ef18616ea6d47458ae6f91f91cdc
(numpy_test)

```

**Comment by mattip** [comment]
@RonaldAJ thanks for confirmation, and thanks for helping track this down.

**Comment by seiko2plus** [comment]
> Has this been reported to MSVC? I tried [seraching their issue forum](https://developercommunity.visualstudio.com/VisualStudio?ftype=problem&q=compiler+reordering+avx256+vmovdqu) (login required?) but perhaps my search foo is not up to the task. A new issue can be reported via https://developercommunity.visualstudio.com/cpp/report , the godbolt link will probably be enough to demonstrate the issue.

I made a report here's the link:
https://developercommunity.visualstudio.com/t/Compiler-generates-unexpected-AVX2-move-/10393417

**Comment by mattip** [comment]
> I made a report

Thanks! I upvoted it :)

**Comment by mattip** [comment]
The MSVC bug has been [fixed internally](https://developercommunity.visualstudio.com/t/Compiler-generates-unexpected-AVX2-move-/10393417) and will be available in the next Visual Studio release

**Comment by mattip** [comment]
A fix for the MSVC bug is now available on the latest version available for download. Should we add a version macro to [the work-around](https://github.com/numpy/numpy/pull/23953/files#diff-dfba9ccd94c7358b0cedef6253028014d5e298f4473e7aebf915d58935afdeb4R375) from #23953?
